### ISSUE TITLE
Polars regex replacement pattern differs from Python - $1 instead of \1

### ISSUE THREAD

**Issue by jbarlow83** [pola-rs/polars#6683]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

Python uses regex replacement strings named \1, \2, etc. to indicate that a matched number group should be replaced. Since this is in the Python standard library, Python users expect other regex implementations to work the same.

Polars regex appears to use $1, $2, instead, like rust. This is different is [not documented](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.str.replace.html#polars.Expr.str.replace).

Perhaps it is worth documenting more broadly that, e.g. Polars uses rust's regular expression engine rather than Python's so one should consult it. It may be worth noting important differences.

### Reproducible example

```python
>>> re.sub(r"(?ix) ^ [(] ([^)]+) [)] $", r"\1", "(unbracket me)")

"unbracket me"

>>> df = pl.DataFrame({"value": ["(unbracket me)"]})
>>> df.with_columns(pl.col("value").str.replace(r"(?ix) ^ [(] ([^)]+) [)] $", r"\1"))

"\1"
```

### Expected behavior

At minimum this needs to be documented, assuming it is designed behavior. 

Consider issuing a warning when the regex replacement string contains strings like `\1` or `\2` and literal=False, since this is probably an error.

Ideally, Polars would follow Python conventions and accept/use \1 as the replacement string by default.

### Installed versions

<details>

```
---Version info---
Polars: 0.16.1
Index type: UInt32
Platform: Linux-6.0.12-76060006-generic-x86_64-with-glibc2.35
Python: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
---Optional dependencies---
pyarrow: 9.0.0
pandas: 1.3.5
numpy: 1.21.5
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: 0.8
deltalake: <not installed>
matplotlib: 3.5.1
```

</details>

**Comment by alexander-beedie** [comment]
Very good point about the documentation - I'll see about improving this today. As you have correctly spotted, the underlying regex engine is operating in Rust so there are necessarily a few small differences and we should better-highlight those 👍

**Comment by ritchie46** [comment]
I agree on documenting the regex engine, I don't agree on parsing the regex strings and giving warnings, nor using python's regex engine.

**Comment by zundertj** [comment]
Changed this to documentation label. This is a similar issue as datetime parsing, where we use chrono. That has been documented everywhere, but for regex it is indeed missing.

**Comment by uurl** [comment]
I opened #27407 to explicitly mention that Python-style replacement backreferences such as `\1` are not supported, but it was closed as unnecessary because the existing docs already explain `$1` / `${1}` syntax.

Given that, should this issue be closed as already documented?

**Comment by orlp** [comment]
@uurl Oops, I didn't see this issue. I consider this solved, we do indeed document `$1` and such.
